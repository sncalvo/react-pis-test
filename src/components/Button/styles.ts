import { BLACK, ERROR_COLOR, PRIMARY, SECONDARY } from '@constants/colors';
import { StyleSheet, ViewStyle } from 'react-native';

interface TextButtonStyle {
  primary: ViewStyle;
  secondary: ViewStyle;
  tertiary: ViewStyle;
  buttonContainer: ViewStyle;
  buttonOutlineBlack: ViewStyle;
  buttonOutlineWhite: ViewStyle;
  buttonFillBlack: ViewStyle;
  buttonFillWhite: ViewStyle;
  buttonDisabled: ViewStyle;
  small: ViewStyle;
}

export default StyleSheet.create<TextButtonStyle>({
  primary: {
    backgroundColor: PRIMARY,
    color: BLACK,
  },
  secondary: {
    backgroundColor: SECONDARY,
    color: BLACK,
  },
  tertiary: {
    backgroundColor: ERROR_COLOR,
    color: BLACK,
  },
  buttonContainer: {
    alignItems: 'center',
    alignSelf: 'stretch',
    borderRadius: 3,
    padding: 10,
  },
  buttonFillBlack: {},
  buttonFillWhite: {},
  buttonOutlineBlack: {
    borderWidth: 1,
  },
  buttonOutlineWhite: {
    borderWidth: 1,
  },
  buttonDisabled: {
    opacity: 0.6,
  },
  small: {
    width: 200,
  },
});
