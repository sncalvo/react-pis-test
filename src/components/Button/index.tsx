import React from 'react';
import { StyleProp, TextStyle, TouchableOpacity, ViewStyle } from 'react-native';

import LoadingIndicator from '@components/LoadingIndicator';

import layoutStyles from '@styles/layout';
import textStyles from '@styles/text';
import styles from './styles';

enum ButtonType {
  Primary,
  Secondary,
  Danger,
}

const BUTTON_STYLES: Map<
  ButtonType,
  {
    container: ViewStyle;
    text: TextStyle;
  }
> = new Map([
  [ButtonType.Primary, { container: styles.primary, text: textStyles.black }],
  [ButtonType.Secondary, { container: styles.secondary, text: textStyles.white }],
  [ButtonType.Danger, { container: styles.tertiary, text: textStyles.white }],
]);

interface TextButtonProps {
  children?: React.ReactChild;
  outline?: boolean;
  type?: ButtonType;
  style?: StyleProp<ViewStyle>;
  onPress?: () => void;
  invert?: boolean;
  disabled?: boolean;
  small?: boolean;
  testID?: string;
  loading?: boolean;
}

const Button: React.FC<TextButtonProps> = ({
  children = '',
  outline = false,
  type = ButtonType.Primary,
  style,
  onPress,
  invert,
  disabled,
  small = false,
  testID,
  loading = false,
}) => {
  const outlineColor = invert ? styles.buttonOutlineWhite : styles.buttonOutlineBlack;
  const fillColor = invert ? styles.buttonFillWhite : styles.buttonFillBlack;
  const buttonStyle = [outline ? outlineColor : fillColor, styles.buttonContainer, style];

  const { container } = BUTTON_STYLES.get(type) ?? {};

  return (
    <TouchableOpacity
      style={[
        buttonStyle,
        disabled && styles.buttonDisabled,
        small && styles.small,
        layoutStyles.rowDirection,
        layoutStyles.alignCenter,
        layoutStyles.justifyCenter,
        container,
      ]}
      onPress={onPress}
      disabled={disabled || loading}
      testID={testID}
    >
      {loading ? <LoadingIndicator /> : children}
    </TouchableOpacity>
  );
};

export { ButtonType };
export default Button;
