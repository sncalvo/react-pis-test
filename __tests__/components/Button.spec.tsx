import React from 'react';

import Button from '@components/Button';

import { fireEvent, render } from '@testing-library/react-native';

import renderer from 'react-test-renderer';

describe('Button', () => {
  it('renders correctly', () => {
    const result = renderer.create(<Button />);
    expect(result).toMatchSnapshot();
  });

  it('will trigger action when pressed', () => {
    const mockOnPress = jest.fn();

    const { getByTestId } = render(<Button onPress={mockOnPress} testID="button" />);

    fireEvent.press(getByTestId('button'));
    expect(mockOnPress).toBeCalled();
  });

  it('can be disabled', () => {
    const mockOnPress = jest.fn();

    const { getByTestId } = render(<Button onPress={mockOnPress} disabled testID="button" />);

    fireEvent.press(getByTestId('button'));
    expect(mockOnPress).not.toBeCalled();
  });
});
