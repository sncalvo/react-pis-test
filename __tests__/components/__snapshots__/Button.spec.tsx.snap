// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#40d2c8",
      "borderRadius": 3,
      "color": "#0F0F0F",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
    }
  }
>
  
</View>
`;
